### PR TITLE
Add Floatplane Presnce

### LINKED ISSUE #1185: Floatplane | www.floatplane.com
**Prerequisites and essential questions** 
* [x] Is it a popular site?
* [x] Is the website older than 2 months? 
* [ ] Is the site locked to a specific country/region?
* [x] Is the site a paid service? (e.g. Netflix, Hulu)
* [ ] Does the website feature NSFW content? (e.g. porn, etc...)
* [ ] Are you a donator/patron?
* [x] Do you acknowledge that coding presences is completely voluntary and may take time for your service to be added regardless of priority?

**What's your Discord username?** 
Tlgenzer#8880
**What's the name of the service?** 
Floatplane | floatplane.com
**What should the Presence display?** 
The video you are watching on the site just like youtube.
**If possible, please provide a logo for the service (512x512 minimum)** 
image

### PR DESCRIPTION
**Pull request by TheGamerzs**
Fixes #1185
image
image
image

### CODE REVIEW

**Review by EncryptedDev — CHANGES_REQUESTED** [review]

**Comment by TheGamerzs** [comment]
image
Stream tested

File: websites/F/Floatplane/presence.ts [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+  browsingTimestamp = new Date().getTime();
+
+presence.on("UpdateData", async () => {
+  const presenceData: PresenceData = {
+      largeImageKey: "logo",
+      startTimestamp: browsingTimestamp
+    },
+    pathnames = location.pathname.split("/");
+
+  if (pathnames[1] === "") {
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:13]: Probably a better check
```suggestion
  if (pathnames[1].length >= 0) {
```

File: websites/F/Floatplane/presence.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  } else if (pathnames[1] === "channel") {
+    //Viewing a channel
+    const channelTitle = document.querySelector(".channel-title")?.textContent;
+
+    //Wait for page to load
+    if (!channelTitle) return;
+
+    presenceData.details = "Viewing channel:";
+    presenceData.state = channelTitle;
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:73]: remove empty line

File: websites/F/Floatplane/presence.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    presenceData.largeImageKey = pathnames[2].toLowerCase();
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:75]: ditto

File: websites/F/Floatplane/presence.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    presenceData.smallImageKey = "logo";
+    presenceData.smallImageText = document.title;
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:78]: ditto

File: websites/F/Floatplane/presence.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+      //Stream
+      const title = document.querySelector(".title-text")?.textContent,
+        video = document.querySelector("video") as HTMLVideoElement;
+
+      if (!title || !video) return;
+
+      presenceData.details = title;
+
+      delete presenceData.startTimestamp;
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:91]: Makes more sense to do it this way as you're assigning values on the line below
```suggestion
      delete presenceData.startTimestamp;
      presenceData.details = title;
```

File: websites/F/Floatplane/presence.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        {
+          label: "View Channel",
+          url: (document.querySelector(".channel-title") as HTMLLinkElement)
+            ?.href
+        }
+      ];
+    }
+  } else if (pathnames[2] === "profile") {
+    //User
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:108]: ditto

File: websites/F/Floatplane/presence.ts [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    const channelTitle = document.querySelector(".channel-title").textContent;
+
+    //Wait for page to load
+    if (!channelTitle) return;
+
+    presenceData.details = "Viewing user:";
+    presenceData.state = channelTitle;
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:116]: ditto

File: websites/F/Floatplane/presence.ts [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    //Wait for page to load
+    if (!video) return;
+
+    const timestamps = presence.getTimestampsfromMedia(video),
+      [, endTS] = timestamps;
+
+    presenceData.details = title;
+    presenceData.state = channel;
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:187]: ditto

File: websites/F/Floatplane/presence.ts [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+    presenceData.largeImageKey = channelURL?.split("/").slice(-1)[0];
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:189]: ditto

File: websites/F/Floatplane/presence.ts [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    delete presenceData.startTimestamp;
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:191]: ditto

File: websites/F/Floatplane/presence.ts [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+    presenceData.endTimestamp = endTS;
+
```
Comment (EncryptedDev) [inline websites/F/Floatplane/presence.ts:193]: ditto

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
remove empty lines

**Review by EncryptedDev — CHANGES_REQUESTED** [review]
upload proof once more, a lot of changes have happened since

**Comment by TheGamerzs** [comment]
image
image
image

**Review by EncryptedDev — DISMISSED** [review]

File: websites/F/Floatplane/dist/metadata.json [hunk 12]
```diff
@@ -0,0 +1,33 @@
+{
+  "$schema": "https://schemas.premid.app/metadata/1.3",
+  "service": "Floatplane",
+  "author": {
+    "name": "The Gamerzs",
+    "id": "730142125181894657"
+  },
+
+  "url": "www.floatplane.com",
```
Comment (Bas950) [inline websites/F/Floatplane/dist/metadata.json:9]:
```suggestion
  },
  "url": "www.floatplane.com",
```

**Review by Bas950 — CHANGES_REQUESTED** [review]

**Review by Bas950 — APPROVED** [review]

**Review by EncryptedDev — APPROVED** [review]

**Pull request merged**